executor/src/witgen/machines/double_sorted_witness_machine.rs

```diff
@@ -1,199 +1,58 @@
-use std::collections::{BTreeMap, HashMap, HashSet};
-use std::iter::once;
-
-use itertools::Itertools;
+use std::collections::HashMap;
 
 use super::{Machine, MachineParts};
 use crate::witgen::rows::RowPair;
-use crate::witgen::util::try_to_simple_poly;
-use crate::witgen::{EvalError, EvalResult, FixedData, MutableState, QueryCallback};
-use crate::witgen::{EvalValue, IncompleteCause};
-
-use powdr_number::{DegreeType, FieldElement};
-
-use powdr_ast::analyzed::{DegreeRange, IdentityKind, PolyID};
-
-/// If all witnesses of a machine have a name in this list (disregarding the namespace),
-/// we'll consider it to be a double-sorted machine.
-/// This does not include the selectors, which are dynamically added to this list.
-const ALLOWED_WITNESSES: [&str; 8] = [
-    "m_value",
-    "m_addr",
-    "m_step",
-    "m_change",
-    "m_is_write",
-    "m_is_bootloader_write",
-    "m_diff_upper",
-    "m_diff_lower",
-];
-
-const DIFF_COLUMNS: [&str; 2] = ["m_diff_upper", "m_diff_lower"];
-const BOOTLOADER_WRITE_COLUMN: &str = "m_is_bootloader_write";
+use crate::witgen::{EvalResult, FixedData, MutableState, QueryCallback};
+use powdr_number::{FieldElement, KnownField};
 
-// The operation ID is decomposed into m_is_write + 2 * m_is_bootloader_write
-const OPERATION_ID_WRITE: u64 = 1;
-const OPERATION_ID_BOOTLOADER_WRITE: u64 = 2;
-
-fn split_column_name(name: &str) -> (&str, &str) {
-    let mut limbs = name.split("::");
-    let namespace = limbs.next().unwrap();
-    let col = limbs.next().unwrap();
-    (namespace, col)
-}
-
-/// TODO make this generic
-
-pub struct DoubleSortedWitnesses<'a, T: FieldElement> {
-    degree_range: DegreeRange,
-    degree: DegreeType,
-    //key_col: String,
-    /// Position of the witness columns in the data.
-    /// The key column has a position of usize::max
-    //witness_positions: HashMap<String, usize>,
-    /// (addr, step) -> value
-    trace: BTreeMap<(T, T), Operation<T>>,
-    /// A map addr -> value, the current content of the memory.
-    data: BTreeMap<T, T>,
-    is_initialized: BTreeMap<T, bool>,
-    namespace: String,
-    name: String,
-    parts: MachineParts<'a, T>,
-    /// If the machine has the `m_diff_upper` and `m_diff_lower` columns, this is the base of the
-    /// two digits.
-    diff_columns_base: Option<u64>,
-    /// Whether this machine has a `m_is_bootloader_write` column.
-    has_bootloader_write_column: bool,
-    /// All selector IDs that are used on the right-hand side connecting identities.
-    selector_ids: BTreeMap<u64, PolyID>,
-}
+use crate::witgen::machines::double_sorted_witness_machine_16;
+use crate::witgen::machines::double_sorted_witness_machine_32;
 
-struct Operation<T> {
-    pub is_normal_write: bool,
-    pub is_bootloader_write: bool,
-    pub value: T,
-    pub selector_id: PolyID,
+pub enum DoubleSortedWitnesses<'a, T: FieldElement> {
```

Comment:
what about this being a trait (that needs to also implement `Machine`) then just dispatching to a `Box<dyn DoubleSortedWitnesses<T>>`?

Reply:
Tried, but the compiler doesn't like it because the trait itself would have functions with generics. The solution would then be to split that trait into multiple traits, or to use the enum approach that's currently used in the PR.